### ISSUE TITLE
Single-point energy of Drude-polarizable system

### ISSUE THREAD

**Issue by leeping** [pandegroup/openmm#905]
Hi everyone,

I'm interested in evaluating the single-point energy of a system with Drude particles, where the Drude particle positions are minimized by SCF.  However, `simulation.context.getState(getEnergy=True)` does not actually optimize the Drude particle positions.  This is slightly different from the behavior in AMOEBA where the energy of the induced dipoles is always calculated before obtaining the energy.  

If I take a miniscule time step (e.g. <0.1 fs) before getting the energy, the Drude particle positions are then properly minimized, but it's still cumbersome (e.g. if I'm trying to get the energy from another program) and not completely correct.  If I'm using the Drude Langevin integrator, then I don't see an easy way to get the energy of the system where the Drude particle positions are optimized.

Just to get peoples' opinions: Is it worthwhile to implement something where the outcome of simulation.context.getState(getEnergy=True) can have minimized Drude particle positions?  If so, where should this be implemented?  (For example, as an extra argument to getState?)  Alternatively, am I being whiny and I should just take the small MD step?  :)

Thanks,

- Lee-Ping

**Comment by swails** [comment]
@leeping -- have you tried setting all particle masses to 0 and *then* taking your small time steps?  I would naively think that this would constrain all atomic positions *and* allow the Drude particles to minimize.  It's still a little clumsy, but should eliminate the errors from the small particle perturbations (and you can wrap that clumsiness into a reusable function).

I'm not sure if @peastman will have counterarguments to this, but my naive thought would be "yes, `getState` should return the *actual* energy of the system" -- if that means minimizing Drude particle locations, then that should be done.

I think Forces have the ability to update particle positions (for example, `MonteCarloBarostat` and friends).

**Comment by leeping** [comment]
@swails , that's a great idea.  I just tried and it works.  :)  With a 1 fs time step, here's what I'm getting:

```
Before time step:
  O  -0.15019999  -0.01910000   0.14350000
 H1  -0.06010000  -0.05970000   0.15539999
 H2  -0.20070000  -0.04220000   0.22200000
  M  -0.14207765  -0.03216547   0.16204189
 OD  -0.15019999  -0.01910000   0.14350000
  O  -0.17450000  -0.03820000  -0.13090000
 H1  -0.18889999  -0.04800000  -0.03480000
 H2  -0.25170001  -0.07670000  -0.17340000
  M  -0.19328803  -0.04810679  -0.11990613
 OD  -0.17450000  -0.03820000  -0.13090000
...
Movement after time step:
  O   0.00000000   0.00000000   0.00000000
 H1   0.00000000   0.00000000   0.00000000
 H2   0.00000000   0.00000000   0.00000000
  M   0.00000000   0.00000000   0.00000000
 OD  -0.00380288   0.00355940  -0.00592180
  O   0.00000000   0.00000000   0.00000000
 H1   0.00000000   0.00000000   0.00000000
 H2   0.00000000   0.00000000   0.00000000
  M   0.00000000   0.00000000   0.00000000
 OD   0.00361016   0.00096461  -0.00289789
...
```

There's still a bit of clumsiness because I need to re-create the Simulation with the zero-mass System.  This needs to happen outside the function that gets the energy, because the energy call is inside an inner loop.  It shouldn't be a problem any more though.

**Comment by leeping** [comment]
I did notice that I can no longer apply constraints with the masses set to zero. This isn't a problem for me, but one to keep in mind.

**Comment by swails** [comment]
> There's still a bit of clumsiness because I need to re-create the Simulation with the zero-mass System. This needs to happen outside the function that gets the energy, because the energy call is inside an inner loop. This is a very reasonable workaround though.

Why did you need to do that?  What happens if you just do `system.setParticleMass()` to change the masses of the same system (avoiding the need to regenerate it)?  I don't think `System` has an `updateParametersInContext` equivalent, so either (a) that won't work or (b) it does work and it's not needed.

> I did notice that I can no longer apply constraints with the masses set to zero. This isn't a problem for me, but one to keep in mind.

Sure you can :).  But it just may not be possible to satisfy all constraints simultaneously (i.e., you can't constrain particle positions *and* satisfy distance constraints not already satisfied by the present conditions simultaneously).  But I guess that's what you meant...

> I'm not sure if @peastman will have counterarguments to this, but my naive thought would be "yes, getState should return the actual energy of the system" -- if that means minimizing Drude particle locations, then that should be done.

I just thought of a counterargument, actually.  I *think* that this is basically what the Drude SCF integrator does, and this is why the Drude Langevin integrator is so much faster (i.e., it *doesn't* need to minimize Drude particle locations).  Of course, I'd be happy to be set straight if I'm misunderstanding something.

**Comment by leeping** [comment]
1) If I set the particle masses to zero after the `Simulation` is created, either using `simulation.system.setParticleMass(i, 0.0)` or `system.setParticleMass(i, 0.0)`, the particles still move.  I think that in order to reset the particle masses, it's necessary to recreate the `Simulation`.

2) That's right; if I had some coordinates and my goal was to first apply the bond / angle / rigid water constraints and then calculate the energy, setting the masses to zero won't work because that adds a position constraint that can't be satisfied at the same time as the others. In practice, the position constraint overrides the other constraints.

3) The question of the "actual" energy of the system, in my opinion, is a bit semantic.  

For all of the Drude-particle force fields that I know about, the Drude particle represents an induced dipole, and the exact energy of the system is given by minimizing the Drude particle locations.  However, the most efficient way to integrate the equations of motion is to use extended Lagrangian methods, and in those methods the Drude particle locations are never minimized - except maybe at the beginning.  

When we use the `DrudeLangevinIntegrator` and call `Simulation.context.getState(getEnergy=True)`, which energy is more reasonable to give the user?  As you mentioned, one could argue it's the energy corresponding to the current particle positions in the simulation, where the Drude particle positions are not minimized (but hopefully close, if the method is working.)

When we're interested in single point energies, it's very natural to minimize the Drude particle positions.  One way to phrase the problem is that the abstract method `solveDipoles()` is being called in different places in the code depending on which polarizable force field we're using; in AMOEBA it's contained inside `context.calcForcesAndEnergy()`, and for the SCF Drude particles isn't.

**Comment by jchodera** [comment]
I agree with @leeping here that for simulations employing the SCF scheme (but not the extended Lagrangian scheme), it *does* make sense to have the energies reported by `context.getState(getEnergy=True)` be the minimized energies.  Maybe an integrator isn't the optimal way to implement the SCF scheme due to this reason?  I could imagine the `DrudeForce` object having `useSCF(bool)` and `setSCFTolerance(tol)` options and handling the SCF solution internally instead, though I know nothing about how hard that implementation change would be or the cost:benefit tradeoff for this.

**Comment by swails** [comment]
@leeping -- I agree with your assessment that `getState(getEnergy=True)` should give you the "correct energy", and follow what the `AmoebaMultipoleForce` does by analogy (i.e., report the energy with the converged induced dipoles).

This is probably more important when we want energies, rather than just forces, but let's see what @peastman has to say about the performance implications of various implementation options, and what it is that we *really* want (i.e., is it enough to just minimize the Drude particles when `getEnergy=True`, do we also want it for `getForce=True` which is typically an MD step we are not reporting energies on, ...)?

**Comment by leeping** [comment]
@swails - I personally think the behavior should be the same with either `getEnergy=True` or `getForce=True`.

**Comment by peastman** [comment]
I think the question of what is meant by "the energy" is more than just semantic.  If you're running a simulation with extended Langrangian, the energy for the current (non-minimized) positions should be viewed as the "true" energy for most purposes.  That's what's determining the dynamics.  If you minimize the positions first, the resulting energies don't correspond to any well defined probability distribution.

Basically, we can talk about three different cases:

1. Run dynamics with extended Lagrangian, record non-minimized energies.
2. Run dynamics with extended Lagrangian, record minimized energies.
3. Run dynamics with SCF, record minimized energies.

As long as the extended Lagrangian scheme is being used appropriately, the Drude particles will always be very close to their minimum energy configuration, and all three of these will produce similar results.  In that case, it doesn't matter which you use.  But to the extent that they produce different results, is 1 or 2 more "correct"?  I'm not really sure about that.  It probably depends on exactly what you plan to use the energies for.

**Comment by leeping** [comment]
I spoke to Peter and we agreed that a nice practical solution would be to add a `SCF()` or `optimizeDrudeParticlePositions()` method to `DrudeSCFIntegrator`.  So it could be called like this:

```
simulation.context.setPositions(modeller.positions)
simulation.integrator.optimizeDrudeParticlePositions(tolerance)
simulation.context.getState(getEnergy=True)
```

If `DrudeLangevinIntegrator` also had this method, that would make things exceedingly easy on my end, but is that a good idea?
